tools/rpkg/src/statement.cpp

```diff
@@ -202,6 +205,24 @@ static SEXP allocate(const LogicalType &type, RProtector &r_varvalue, idx_t nrow
 	case LogicalTypeId::LIST:
 		varvalue = r_varvalue.Protect(NEW_LIST(nrows));
 		break;
+	case LogicalTypeId::STRUCT: {
+		cpp11::writable::list dest_list;
+
+		for (const auto &child : StructType::GetChildTypes(type)) {
+			const auto &name = child.first;
+			const auto &child_type = child.second;
+
+			RProtector child_protector;
+			auto dest_child = allocate(child_type, child_protector, nrows);
+			dest_list.push_back(cpp11::named_arg(name.c_str()) = std::move(dest_child));
+		}
+
+		dest_list.attr(R_ClassSymbol) = RStrings::get().dataframe_str;
+		dest_list.attr(R_RowNamesSymbol) = {NA_INTEGER, -static_cast<int>(nrows)};
```

Comment:
Is there a more idiomatic way to return a data frame with cpp11?

Reply:
This is intentional - will add a comment to explain. cpp11's [data frame constructor](https://github.com/r-lib/cpp11/blob/322e5ee67346926993cc6f464152802118250b78/inst/include/cpp11/data_frame.hpp#L70) wouldn't work here as it tries to calculate the number of rows itself, and gives the wrong answer if the first column is itself another data frame or if the struct is empty.